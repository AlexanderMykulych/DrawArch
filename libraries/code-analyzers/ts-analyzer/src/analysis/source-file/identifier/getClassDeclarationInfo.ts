import type { ClassDeclaration, MethodDeclaration } from "ts-morph";
import { EntityType, FunctionEntity, CodeInfoType, BaseEntity } from "../../types";
import { getLocation } from "./getLocation";
import { getParentId, getParentsInfo } from "./getParentId";

export function getClassDeclarationInfo(node: ClassDeclaration): BaseEntity {
  const sourceFile = node.getSourceFile()
  return {
    id: `${getParentId(node)}/${node.getName()}`,
    entityType: EntityType.Class,
    type: CodeInfoType.Entity,
    filePath: sourceFile.getFilePath(),
    ...getLocation(node),
  }
}