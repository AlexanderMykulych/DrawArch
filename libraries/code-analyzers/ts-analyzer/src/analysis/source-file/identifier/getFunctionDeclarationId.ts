import type { FunctionDeclaration } from "ts-morph";
import { CodeInfoType, EntityType, FunctionEntity } from "../../types";
import { getLocation } from "./getLocation";
import { getParentId, getParentsInfo } from "./getParentId";

export function getFunctionDeclarationId(fnDec: FunctionDeclaration): FunctionEntity {
  const sourceFile = fnDec.getSourceFile()

  return {
    id: `${getParentId(fnDec)}/${fnDec.getName()}`,
    entityType: EntityType.Function,
    filePath: sourceFile.getFilePath(),
    type: CodeInfoType.Entity,
    typeString: fnDec.getType().getText(),
    source: sourceFile
      ? getParentsInfo(sourceFile)
      : [],
    ...getLocation(fnDec),
  }
}