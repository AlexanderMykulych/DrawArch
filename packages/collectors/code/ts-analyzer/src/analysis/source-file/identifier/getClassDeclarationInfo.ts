import type { ClassDeclaration } from "ts-morph";
import { EntityType, CodeInfoType, BaseEntity } from "../../types";
import { getLocation } from "./getLocation";
import { getParentId } from "./getParentId";

export function getClassDeclarationInfo(node: ClassDeclaration): BaseEntity {
  const sourceFile = node.getSourceFile()
  return {
    id: `${getParentId(node)}/${node.getName()}`,
    entityType: EntityType.Class,
    type: CodeInfoType.Entity,
    filePath: sourceFile.getFilePath(),
    ...getLocation(node),
  }
}