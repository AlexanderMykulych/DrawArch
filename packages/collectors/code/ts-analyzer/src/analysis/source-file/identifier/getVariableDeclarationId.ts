import type { VariableDeclaration } from "ts-morph";
import { EntityType, BaseEntity, CodeInfoType } from "../../types";
import { getLocation } from "./getLocation";
import { getParentId, getParentsInfo } from "./getParentId";

export function getVariableDeclarationId(v: VariableDeclaration): BaseEntity {
  const sourceFile = v.getSourceFile()
  return {
    id: `${getParentId(v)}/${v.getName()}`,
    entityType: EntityType.Variable,
    filePath: sourceFile.getFilePath(),
    type: CodeInfoType.Entity,
    // source: sourceFile
    //   ? getParentsInfo(sourceFile)
    //   : [],
    ...getLocation(v),
  }
}