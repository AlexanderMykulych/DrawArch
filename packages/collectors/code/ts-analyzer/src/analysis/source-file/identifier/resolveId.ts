export function resolveId(...args: string[]) {
  return args.join('/')
}