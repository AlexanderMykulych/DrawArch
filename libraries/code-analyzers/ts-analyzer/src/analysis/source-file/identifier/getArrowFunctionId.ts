import { ArrowFunction } from "ts-morph";
import { EntityType, FunctionEntity, CodeInfoType } from "../../types";
import { getNodeInfo } from "./getNodeId";
import { getParentId } from "./getParentId";

export function getArrowFunctionId(arrowFn: ArrowFunction): FunctionEntity {
  const sourceFile = arrowFn.getSourceFile()
  return {
    id: getParentId(arrowFn),
    entityType: EntityType.Function,
    typeString: arrowFn.getType().getText(),
    type: CodeInfoType.Entity,
    filePath: sourceFile.getFilePath(),
    source: getNodeInfo(sourceFile) ?? '<unknown>'
  }
}