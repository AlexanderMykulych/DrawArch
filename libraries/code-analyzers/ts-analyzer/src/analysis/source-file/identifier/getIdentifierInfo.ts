import { Identifier, Node, SyntaxKind } from "ts-morph";
import { CodeInfoType, Entity, EntityType, RelationType } from "../../types";
import { getRelationBeetwenNodes } from "../relations/getRelationBeetwenNodes";
import { getLocation } from "./getLocation";
import { getNodeInfo } from "./getNodeId";
import { getParentId, getParentEntity, getParentNode } from "./getParentId";
import type { NodeCodeInfos } from "./types";

export function getIdentifierInfo(node: Identifier): NodeCodeInfos {
  const parentNode = getParentNode(node)
  const declarationNode = getDeclarationNode(node)
  
  const isParentNode = parentNode === declarationNode

  if (isParentNode) {
    return getNodeInfo(parentNode)
  }

  return getIdentifierEntity(node)
}

function getIdentifierEntity(node: Identifier): NodeCodeInfos {
  const implementations = node.getImplementations()

  let codeInfos: Entity[] = []
  if (implementations.length > 0) {
    codeInfos = implementations
      .map(x => x.getNode())
      .filter(x => x !== node)
      .flatMap(x => getNodeInfo(x)[0])
  }

  const parentEntity = getParentEntity(node)

  if (codeInfos.length > 0) {

    // @ts-ignore
    return [
      ...codeInfos,
      ...codeInfos
        .map(codeInfo => getRelationBeetwenNodes({
            from: parentEntity,
            to: codeInfo,
            type: RelationType.Use,
          })
        )
    ]
  }

  const declarationNode = node.getSymbol()?.getValueDeclaration() ?? node

  if (!declarationNode.isKind(SyntaxKind.Identifier)) {
    const declarationEntity = getNodeInfo(declarationNode)[0]

    return [
      declarationEntity,
      ...codeInfos,
      getRelationBeetwenNodes({
        from: parentEntity,
        to: declarationEntity,
        type: RelationType.Parent,
        location: getLocation(node),
      }),
    ]
  }

  const location = getLocation(node)

  const identifierEntity: Entity = {
    type: CodeInfoType.Entity,
    entityType: EntityType.Identifier,
    id: `${getParentId(node)}/${node.getText()}`,
    filePath: node.getSourceFile().getFilePath(),
    ...location,
  }

  return [
    identifierEntity
  ]
}

function getDeclarationNode(node: Identifier): Node | undefined {
  return node.getSymbol()?.getValueDeclaration()
}