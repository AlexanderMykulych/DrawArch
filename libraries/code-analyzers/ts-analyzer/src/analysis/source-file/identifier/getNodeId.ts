import { Node, SyntaxKind } from 'ts-morph'
import { Entity, RelationType } from '../../types'
import { getRelationBeetwenNodes } from '../relations/getRelationBeetwenNodes'
import type { AnalyzerOptions } from '../types'
import { getArrowFunctionId } from './getArrowFunctionId'
import { getClassDeclarationInfo } from './getClassDeclarationInfo'
import { getFunctionDeclarationId } from './getFunctionDeclarationId'
import { getIdentifierInfo } from './getIdentifierInfo'
import { getIgnoredNode } from './getIgnoredNode'
import { getImportDeclarationId } from './getImportDeclarationId'
import { getImportSpecifierId } from './getImportSpecifierId'
import { getLocation } from './getLocation'
import { getMethodDeclarationId } from './getMethodDeclarationId'
import { getMethodSignatureIndo } from './getMethodSignatureIndo'
import { getParentEntity } from './getParentId'
import { getPropertyAssignment } from './getPropertyAssignment'
import { getSourceFileId } from './getSourceFileId'
import { getVariableDeclarationId } from './getVariableDeclarationId'
import { isNodeCodeInfos, NodeCodeInfos } from './types'

type NodeExtractor = {
  kind: SyntaxKind
  fn: (node: any) => Entity | NodeCodeInfos
}

const nodeInfoExtractors: NodeExtractor[] = [
  {
    kind: SyntaxKind.ImportSpecifier,
    fn: getImportSpecifierId,
  },
  {
    kind: SyntaxKind.FunctionDeclaration,
    fn: getFunctionDeclarationId,
  },
  {
    kind: SyntaxKind.Identifier,
    fn: getIdentifierInfo,
  },
  {
    kind: SyntaxKind.VariableDeclaration,
    fn: getVariableDeclarationId,
  },
  {
    kind: SyntaxKind.ArrowFunction,
    fn: getArrowFunctionId,
  },
  {
    kind: SyntaxKind.MethodDeclaration,
    fn: getMethodDeclarationId,
  },
  {
    kind: SyntaxKind.SourceFile,
    fn: getSourceFileId,
  },
  {
    kind: SyntaxKind.ImportDeclaration,
    fn: getImportDeclarationId,
  },
  {
    kind: SyntaxKind.PropertyAssignment,
    fn: getPropertyAssignment,
  },
  {
    kind: SyntaxKind.MethodSignature,
    fn: getMethodSignatureIndo,
  },
  {
    kind: SyntaxKind.ClassDeclaration,
    fn: getClassDeclarationInfo,
  },
]

function _getNodeInfo(node: Node): Entity | NodeCodeInfos {
  if (!node) {
    throw '<unknow node>'
  }
  const extractor = nodeInfoExtractors.find(x => node.isKind(x.kind))
  if (extractor) {
    if (!extractor.fn) {
      throw new Error(`${node.getText()}, ${node.getKindName()} ${JSON.stringify(extractor)}`)
    }
    return extractor.fn(node)
  }

  return getIgnoredNode(node)
}


export function getNodeInfo(node: Node, options?: AnalyzerOptions): NodeCodeInfos {
  let result = _getNodeInfo(node)

  if (!isNodeCodeInfos(result)) {
    if (node.isKind(SyntaxKind.SourceFile)) {
      result = [result]
    } else {
      const parent = getParentEntity(node)

      result = [
        result,
        getRelationBeetwenNodes({
          from: parent,
          to: result,
          type: RelationType.Parent,
          location: getLocation(node),
        }),
      ]
    }
  }

  options?.context?.addCodeInfo?.(result)

  return result
}

export function canAnalyzeNode(node: Node): boolean {
  return nodeInfoExtractors.some(x => node.isKind(x.kind))
}