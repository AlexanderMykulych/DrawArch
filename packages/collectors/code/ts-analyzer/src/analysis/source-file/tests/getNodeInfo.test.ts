import { Project, SyntaxKind } from "ts-morph";
import { test, expect } from "vitest";
import { getNodeInfo } from "../identifier/getNodeId";

test('should analyze', () => {
  const project = new Project({
    useInMemoryFileSystem: true,
  });
  const utilsSource = project.createSourceFile('utils.ts', 'export function sum(a: number, b: number) { return a + b }"')
  const fn = utilsSource.getDescendantsOfKind(SyntaxKind.FunctionDeclaration)[0]
  const result = fn ? getNodeInfo(fn) : null

  expect(result).toMatchSnapshot(result)
})
