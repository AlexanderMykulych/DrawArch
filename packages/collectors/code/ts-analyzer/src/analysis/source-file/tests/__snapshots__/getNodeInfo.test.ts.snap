// Vitest Snapshot v1

exports[`should analyze 1`] = `
[
  {
    "code": " { return a + b }",
    "columnEnd": 59,
    "columnStart": 1,
    "end": 58,
    "entityType": "function",
    "filePath": "/utils.ts",
    "id": "/utils.ts/sum",
    "kind": "FunctionDeclaration",
    "lineEnd": 1,
    "lineStart": 1,
    "logEnd": 58,
    "logStart": 42,
    "start": 0,
    "type": "entity",
    "typeString": "typeof import(\\"/utils\\").sum",
  },
  {
    "code": " { return a + b }",
    "columnEnd": 59,
    "columnStart": 1,
    "end": 58,
    "from": "/utils.ts",
    "id": "/utils.ts->/utils.ts/sum",
    "kind": "FunctionDeclaration",
    "lineEnd": 1,
    "lineStart": 1,
    "logEnd": 58,
    "logStart": 42,
    "relationType": "parent",
    "start": 0,
    "to": "/utils.ts/sum",
    "type": "relationship",
  },
]
`;
